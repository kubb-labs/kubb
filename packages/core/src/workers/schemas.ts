import { z } from 'zod'

import type { ZodStringDef, ZodType } from 'zod'
import type { KubbFile } from '../FileManager.ts'

export const importSchema = z.object({
  name: z.union([z.string(), z.array(z.string())]),
  path: z.string(),
  isTypeOnly: z.boolean().optional(),
  isNameSpace: z.boolean().optional(),
  root: z.string().optional(),
})

export const exportSchema = z.object({
  name: z.union([z.string(), z.array(z.string())]).optional(),
  path: z.string(),
  isTypeOnly: z.boolean().optional(),
  asAlias: z.boolean().optional(),
})

export const fileSchema = z.object({
  id: z.string().optional(),
  baseName: z.string() as ZodType<KubbFile.BaseName, ZodStringDef>,
  path: z.string(),
  source: z.string(),
  imports: z.array(importSchema).optional(),
  exports: z.array(exportSchema).optional(),
  override: z.boolean().optional(),
  meta: z.object({}).optional(),
  env: z.object({}).optional(),
})
