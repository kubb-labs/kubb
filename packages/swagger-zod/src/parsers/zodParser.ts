export const zodKeywords = {
  any: 'any',
  number: 'number',
  integer: 'integer',
  object: 'object',
  lazy: 'lazy',
  string: 'string',
  boolean: 'boolean',
  undefined: 'undefined',
  null: 'null',
  nullish: 'nullish',
  array: 'array',
  tuple: 'tuple',
  enum: 'enum',
  union: 'union',
  datetime: 'datetime',
  email: 'email',
  uuid: 'uuid',
  url: 'url',
  /* intersection */
  default: 'default',
  and: 'and',
  describe: 'describe',
  min: 'min',
  max: 'max',
  optional: 'optional',
  catchall: 'catchall',

  // custom ones
  ref: 'ref',
  matches: 'matches',
} as const

export type ZodKeyword = keyof typeof zodKeywords

export const zodKeywordMapper: Record<ZodKeyword, string> = {
  any: 'z.any',
  number: 'z.number',
  integer: 'z.number',
  object: 'z.object',
  lazy: 'z.lazy',
  string: 'z.string',
  boolean: 'z.boolean',
  undefined: 'z.undefined',
  null: '.nullable',
  nullish: '.nullish',
  array: 'z.array',
  tuple: 'z.tuple',
  enum: 'z.enum',
  union: 'z.union',
  datetime: '.datetime',
  email: '.email',
  uuid: '.uuid',
  url: '.url',
  /* intersection */
  default: '.default',
  and: '.and',
  describe: '.describe',
  min: '.min',
  max: '.max',
  optional: '.optional',
  catchall: '.catchall',

  // custom ones
  ref: 'ref',
  matches: '.regex',
} as const

type ZodMetaBase<T> = {
  keyword: ZodKeyword
  args: T
}

type ZodMetaAny = { keyword: typeof zodKeywords.any }
type ZodMetaNull = { keyword: typeof zodKeywords.null }

type ZodMetaNullish = { keyword: typeof zodKeywords.nullish }
type ZodMetaUndefined = { keyword: typeof zodKeywords.undefined }

type ZodMetaNumber = { keyword: typeof zodKeywords.number }
type ZodMetaInteger = { keyword: typeof zodKeywords.integer }

type ZodMetaString = { keyword: typeof zodKeywords.string }

type ZodMetaBoolean = { keyword: typeof zodKeywords.boolean }

type ZodMetaDescribe = { keyword: typeof zodKeywords.describe; args?: string }
type ZodMetaMin = { keyword: typeof zodKeywords.min; args?: number }

type ZodMetaMax = { keyword: typeof zodKeywords.max; args?: number }
type ZodMetaMatches = { keyword: typeof zodKeywords.matches; args?: string }
type ZodMetaOptional = { keyword: typeof zodKeywords.optional }

type ZodMetaObject = { keyword: typeof zodKeywords.object; args?: { [x: string]: ZodMeta[] } }

type ZodMetaCatchall = { keyword: typeof zodKeywords.catchall; args?: ZodMeta[] }

type ZodMetaRef = { keyword: typeof zodKeywords.ref; args?: string }

type ZodMetaUnion = { keyword: typeof zodKeywords.union; args?: ZodMeta[] }

type ZodMetaAnd = { keyword: typeof zodKeywords.and; args?: ZodMeta[] }

type ZodMetaEnum = { keyword: typeof zodKeywords.enum; args?: Array<string | number> }

type ZodMetaArray = { keyword: typeof zodKeywords.array; args?: ZodMeta[] }

type ZodMetaTuple = { keyword: typeof zodKeywords.tuple; args?: ZodMeta[] }
type ZodMetaLazy = { keyword: typeof zodKeywords.lazy }
type ZodMetaDefault = { keyword: typeof zodKeywords.default; args?: string | number | boolean }

type ZodMetaDatetime = { keyword: typeof zodKeywords.datetime }

type ZodMetaEmail = { keyword: typeof zodKeywords.email }

type ZodMetaUuid = { keyword: typeof zodKeywords.uuid }

type ZodMetaUrl = { keyword: typeof zodKeywords.url }

export type ZodMeta =
  | ZodMetaAny
  | ZodMetaNull
  | ZodMetaNullish
  | ZodMetaUndefined
  | ZodMetaInteger
  | ZodMetaNumber
  | ZodMetaString
  | ZodMetaBoolean
  | ZodMetaLazy
  | ZodMetaDescribe
  | ZodMetaMin
  | ZodMetaMax
  | ZodMetaMatches
  | ZodMetaOptional
  | ZodMetaObject
  | ZodMetaCatchall
  | ZodMetaRef
  | ZodMetaUnion
  | ZodMetaAnd
  | ZodMetaEnum
  | ZodMetaArray
  | ZodMetaTuple
  | ZodMetaDefault
  | ZodMetaDatetime
  | ZodMetaEmail
  | ZodMetaUuid
  | ZodMetaUrl

/**
 *
 * @link based on https://github.com/cellular/oazapfts/blob/7ba226ebb15374e8483cc53e7532f1663179a22c/src/codegen/generate.ts#L398
 */

function zodKeywordSorter(a: ZodMeta, b: ZodMeta) {
  if (b.keyword === zodKeywords.null) {
    return -1
  }

  return 0
}

export function parseZodMeta(item: ZodMeta): string {
  // eslint-disable-next-line prefer-const
  let { keyword, args = '' } = (item || {}) as ZodMetaBase<any>
  const value = zodKeywordMapper[keyword]

  if (keyword === zodKeywords.tuple) {
    return `${value}(${Array.isArray(args) ? `[${args.map(parseZodMeta).join(',')}]` : parseZodMeta(args as ZodMeta)})`
  }

  if (keyword === zodKeywords.array) {
    return `${value}(${Array.isArray(args) ? `${args.map(parseZodMeta).join('')}` : parseZodMeta(args as ZodMeta)})`
  }
  if (keyword === zodKeywords.union) {
    return `${Array.isArray(args) ? `${value}([${args.map(parseZodMeta).join(',')}])` : parseZodMeta(args as ZodMeta)}`
  }

  if (keyword === zodKeywords.catchall) {
    return `${value}(${Array.isArray(args) ? `${args.map(parseZodMeta).join('')}` : parseZodMeta(args as ZodMeta)})`
  }

  if (keyword === zodKeywords.and && Array.isArray(args)) {
    return `${args
      .map(parseZodMeta)
      .map((item) => `${value}(${item})`)
      .join('')}`
  }

  if (keyword === zodKeywords.object) {
    if (!args) {
      args = '{}'
    }
    const argsObject = Object.entries(args as ZodMeta)
      .filter((item) => {
        const schema = item[1] as ZodMeta[]
        return schema && typeof schema.map === 'function'
      })
      .map((item) => {
        const key = item[0]
        const schema = item[1] as ZodMeta[]
        return `"${key}": ${schema.sort(zodKeywordSorter).map(parseZodMeta).join('')}`
      })
      .join(',')

    args = `{${argsObject}}`
  }

  // custom type
  if (keyword === zodKeywords.ref) {
    // use of z.lazy because we need to import from files x or we use the type as a self reference
    return `${zodKeywordMapper.lazy}(() => ${args})`
  }

  if (keyword === zodKeywords.default && args === undefined) {
    return ''
  }

  if (keyword in zodKeywords) {
    return `${value}(${args})`
  }

  return '""'
}

export function zodParser(items: ZodMeta[], options: { name: string }): string {
  if (!items.length) {
    return `export const ${options.name} = '';`
  }

  return `export const ${options.name} = ${items.map(parseZodMeta).join('')};`
}
