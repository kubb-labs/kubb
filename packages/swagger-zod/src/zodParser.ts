export const zodKeywords = {
  any: 'any',
  number: 'number',
  integer: 'integer',
  object: 'object',
  lazy: 'lazy',
  string: 'string',
  boolean: 'boolean',
  undefined: 'undefined',
  nullable: 'nullable',
  null: 'null',
  nullish: 'nullish',
  array: 'array',
  tuple: 'tuple',
  enum: 'enum',
  union: 'union',
  literal: 'literal',
  datetime: 'datetime',
  email: 'email',
  uuid: 'uuid',
  url: 'url',
  /* intersection */
  default: 'default',
  and: 'and',
  describe: 'describe',
  min: 'min',
  max: 'max',
  optional: 'optional',
  catchall: 'catchall',
  readOnly: 'readOnly',

  // custom ones
  ref: 'ref',
  matches: 'matches',
} as const

export type ZodKeyword = keyof typeof zodKeywords

export const zodKeywordMapper: Record<ZodKeyword, string> = {
  any: 'z.any',
  number: 'z.number',
  integer: 'z.number',
  object: 'z.object',
  lazy: 'z.lazy',
  string: 'z.string',
  boolean: 'z.boolean',
  undefined: 'z.undefined',
  nullable: '.nullable',
  null: 'z.null',
  nullish: '.nullish',
  array: 'z.array',
  tuple: 'z.tuple',
  enum: 'z.enum',
  union: 'z.union',
  literal: 'z.literal',
  datetime: '.datetime',
  email: '.email',
  uuid: '.uuid',
  url: '.url',
  /* intersection */
  default: '.default',
  and: '.and',
  describe: '.describe',
  min: '.min',
  max: '.max',
  optional: '.optional',
  catchall: '.catchall',
  readOnly: '.readonly',

  // custom ones
  ref: 'ref',
  matches: '.regex',
} as const

type ZodMetaBase<T> = {
  keyword: ZodKeyword
  args: T
}

type ZodMetaUnknown = { keyword: string }

type ZodMetaAny = { keyword: typeof zodKeywords.any }
type ZodMetaNull = { keyword: typeof zodKeywords.null }

type ZodMetaNullish = { keyword: typeof zodKeywords.nullish }
type ZodMetaUndefined = { keyword: typeof zodKeywords.undefined }

type ZodMetaNumber = { keyword: typeof zodKeywords.number }
type ZodMetaInteger = { keyword: typeof zodKeywords.integer }

type ZodMetaString = { keyword: typeof zodKeywords.string }

type ZodMetaBoolean = { keyword: typeof zodKeywords.boolean }

type ZodMetaDescribe = { keyword: typeof zodKeywords.describe; args?: string }
type ZodMetaMin = { keyword: typeof zodKeywords.min; args?: number }

type ZodMetaMax = { keyword: typeof zodKeywords.max; args?: number }
type ZodMetaMatches = { keyword: typeof zodKeywords.matches; args?: string }
type ZodMetaOptional = { keyword: typeof zodKeywords.optional }

type ZodMetaObject = { keyword: typeof zodKeywords.object; args?: { [x: string]: ZodMeta[] } }

type ZodMetaCatchall = { keyword: typeof zodKeywords.catchall; args?: ZodMeta[] }

/**
 * If external, `.schema` will be added
 */
type ZodMetaRef = { keyword: typeof zodKeywords.ref; args?: { name: string; external?: boolean } }

type ZodMetaUnion = { keyword: typeof zodKeywords.union; args?: ZodMeta[] }
type ZodMetaLiteral = { keyword: typeof zodKeywords.literal; args: string | number }

type ZodMetaAnd = { keyword: typeof zodKeywords.and; args?: ZodMeta[] }

type ZodMetaEnum = { keyword: typeof zodKeywords.enum; args?: Array<string | number> }

type ZodMetaArray = { keyword: typeof zodKeywords.array; args?: ZodMeta[] }

type ZodMetaTuple = { keyword: typeof zodKeywords.tuple; args?: ZodMeta[] }
type ZodMetaLazy = { keyword: typeof zodKeywords.lazy }
type ZodMetaDefault = { keyword: typeof zodKeywords.default; args?: string | number | boolean }

type ZodMetaDatetime = { keyword: typeof zodKeywords.datetime }

type ZodMetaEmail = { keyword: typeof zodKeywords.email }

type ZodMetaUuid = { keyword: typeof zodKeywords.uuid }

type ZodMetaUrl = { keyword: typeof zodKeywords.url }
type ZodMetaReadOnly = { keyword: typeof zodKeywords.readOnly }

export type ZodMeta =
  | ZodMetaUnknown
  | ZodMetaAny
  | ZodMetaNull
  | ZodMetaNullish
  | ZodMetaUndefined
  | ZodMetaInteger
  | ZodMetaNumber
  | ZodMetaString
  | ZodMetaBoolean
  | ZodMetaLazy
  | ZodMetaDescribe
  | ZodMetaMin
  | ZodMetaMax
  | ZodMetaMatches
  | ZodMetaOptional
  | ZodMetaObject
  | ZodMetaCatchall
  | ZodMetaRef
  | ZodMetaUnion
  | ZodMetaAnd
  | ZodMetaEnum
  | ZodMetaArray
  | ZodMetaTuple
  | ZodMetaDefault
  | ZodMetaDatetime
  | ZodMetaEmail
  | ZodMetaUuid
  | ZodMetaLiteral
  | ZodMetaUrl
  | ZodMetaReadOnly

/**
 * @link based on https://github.com/cellular/oazapfts/blob/7ba226ebb15374e8483cc53e7532f1663179a22c/src/codegen/generate.ts#L398
 */

function zodKeywordSorter(a: ZodMeta, b: ZodMeta): 1 | -1 | 0 {
  if (b.keyword === zodKeywords.null) {
    return -1
  }

  return 0
}

export function parseZodMeta(item: ZodMeta, mapper: Record<ZodKeyword, string> = zodKeywordMapper): string {
  // eslint-disable-next-line prefer-const
  let { keyword, args = '' } = (item || {}) as ZodMetaBase<unknown>
  const value = mapper[keyword]

  if (keyword === zodKeywords.tuple) {
    return `${value}(${Array.isArray(args) ? `[${args.map((item) => parseZodMeta(item as ZodMeta, mapper)).join(',')}]` : parseZodMeta(args as ZodMeta)})`
  }

  if (keyword === zodKeywords.enum) {
    return `${value}(${Array.isArray(args) ? `[${args.join(',')}]` : parseZodMeta(args as ZodMeta)})`
  }

  if (keyword === zodKeywords.array) {
    return `${value}(${Array.isArray(args) ? `${args.map((item) => parseZodMeta(item as ZodMeta, mapper)).join('')}` : parseZodMeta(args as ZodMeta)})`
  }
  if (keyword === zodKeywords.union) {
    // zod union type needs at least 2 items
    if (Array.isArray(args) && args.length === 1) {
      return parseZodMeta(args[0] as ZodMeta)
    }
    if (Array.isArray(args) && !args.length) {
      return ''
    }

    return `${Array.isArray(args) ? `${value}([${args.map((item) => parseZodMeta(item as ZodMeta, mapper)).join(',')}])` : parseZodMeta(args as ZodMeta)}`
  }

  if (keyword === zodKeywords.catchall) {
    return `${value}(${Array.isArray(args) ? `${args.map((item) => parseZodMeta(item as ZodMeta, mapper)).join('')}` : parseZodMeta(args as ZodMeta)})`
  }

  if (keyword === zodKeywords.and && Array.isArray(args)) {
    return `${
      args
        .map((item) => parseZodMeta(item as ZodMeta, mapper))
        .filter(Boolean)
        .map((item, index) => (index === 0 ? item : `${value}(${item})`))
        .join('')
    }`
  }

  if (keyword === zodKeywords.object) {
    if (!args) {
      args = '{}'
    }
    const argsObject = Object.entries(args as ZodMeta)
      .filter((item) => {
        const schema = item[1] as ZodMeta[]
        return schema && typeof schema.map === 'function'
      })
      .map((item) => {
        const name = item[0]
        const schema = item[1] as ZodMeta[]
        return `"${name}": ${
          schema
            .sort(zodKeywordSorter)
            .map((item) => parseZodMeta(item, mapper))
            .join('')
        }`
      })
      .join(',')

    args = `{${argsObject}}`
  }

  // custom type
  if (keyword === zodKeywords.ref) {
    // use of z.lazy because we need to import from files x or we use the type as a self reference, external will add `.schema`
    const refArgs = args as ZodMetaRef['args']

    if (refArgs?.external) {
      return `${mapper.lazy}(() => ${refArgs?.name}).schema`
    }

    return `${mapper.lazy}(() => ${refArgs?.name})`
  }

  if (keyword === zodKeywords.default && args === undefined) {
    return ''
  }

  if (keyword in mapper) {
    return `${value}(${args as string})`
  }

  return '""'
}

export function zodParser(items: ZodMeta[], options: { keysToOmit?: string[]; mapper?: Record<ZodKeyword, string>; name: string }): string {
  if (!items.length) {
    return `export const ${options.name} = '';`
  }

  if (options.keysToOmit?.length) {
    const omitText = `.omit({ ${options.keysToOmit.map((key) => `${key}: true`).join(',')} })`
    return `export const ${options.name} = ${items.map((item) => parseZodMeta(item, { ...zodKeywordMapper, ...options.mapper })).join('')}${omitText};`
  }

  return `export const ${options.name} = ${items.map((item) => parseZodMeta(item, { ...zodKeywordMapper, ...options.mapper })).join('')};`
}
