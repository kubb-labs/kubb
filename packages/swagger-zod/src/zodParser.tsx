import transformers, { createJSDocBlockText } from '@kubb/core/transformers'
import { isKeyword, schemaKeywords } from '@kubb/swagger'

import type { Schema, SchemaKeywordBase, SchemaMapper } from '@kubb/swagger'

export const zodKeywordMapper = {
  any: 'z.any',
  unknown: 'z.unknown',
  number: 'z.number',
  integer: 'z.number',
  object: 'z.object',
  lazy: 'z.lazy',
  string: 'z.string',
  boolean: 'z.boolean',
  undefined: 'z.undefined',
  nullable: '.nullable',
  null: 'z.null',
  nullish: '.nullish',
  array: 'z.array',
  tuple: 'z.tuple',
  enum: 'z.enum',
  union: 'z.union',
  const: 'z.literal',
  datetime: 'z.string().datetime',
  date: 'z.date',
  uuid: '.uuid',
  url: '.url',
  strict: '.strict',
  default: '.default',
  and: '.and',
  describe: '.describe',
  min: '.min',
  max: '.max',
  optional: '.optional',
  matches: '.regex',
  email: '.email',
  firstName: undefined,
  lastName: undefined,
  password: undefined,
  phone: undefined,
  readOnly: undefined,
  ref: 'ref',
  blob: undefined,
  deprecated: undefined,
  example: undefined,
  type: undefined,
  format: undefined,
  catchall: '.catchall',
} satisfies SchemaMapper

/**
 * @link based on https://github.com/cellular/oazapfts/blob/7ba226ebb15374e8483cc53e7532f1663179a22c/src/codegen/generate.ts#L398
 */

function sort(items?: Schema[]): Schema[] {
  const order: string[] = [
    schemaKeywords.string,
    schemaKeywords.number,
    schemaKeywords.object,
    schemaKeywords.url,
    schemaKeywords.email,
    schemaKeywords.firstName,
    schemaKeywords.lastName,
    schemaKeywords.password,
    schemaKeywords.matches,
    schemaKeywords.uuid,
    schemaKeywords.min,
    schemaKeywords.max,
    schemaKeywords.default,
    schemaKeywords.describe,
    schemaKeywords.optional,
    schemaKeywords.nullable,
    schemaKeywords.nullish,
    schemaKeywords.null,
  ]

  if (!items) {
    return []
  }

  return transformers.orderBy(items, [(v) => order.indexOf(v.keyword)], ['asc'])
}

type ParserOptions = {
  name: string
  typeName?: string
  description?: string

  keysToOmit?: string[]
  mapper?: typeof zodKeywordMapper
}

export function parseZodMeta(item: Schema, options: ParserOptions): string | undefined {
  const mapper = options.mapper || zodKeywordMapper
  const value = mapper[item.keyword as keyof typeof mapper]

  if (!value) {
    return undefined
  }

  if (isKeyword(item, schemaKeywords.union)) {
    // zod union type needs at least 2 items
    if (Array.isArray(item.args) && item.args.length === 1) {
      return parseZodMeta(item.args[0] as Schema, options)
    }
    if (Array.isArray(item.args) && !item.args.length) {
      return ''
    }

    return `${value}([${sort(item.args)
      .map((unionItem) => parseZodMeta(unionItem, options))
      .filter(Boolean)
      .join(', ')}])`
  }

  if (isKeyword(item, schemaKeywords.and)) {
    return sort(item.args)
      .filter((item: Schema) => {
        return ![schemaKeywords.optional, schemaKeywords.describe].includes(item.keyword as typeof schemaKeywords.describe)
      })
      .map((item: Schema) => parseZodMeta(item, options))
      .filter(Boolean)
      .map((item, index) => (index === 0 ? item : `${value}(${item})`))
      .join('')
  }

  if (isKeyword(item, schemaKeywords.array)) {
    return [
      `${value}(${sort(item.args.items)
        .map((arrayItem) => parseZodMeta(arrayItem, options))
        .filter(Boolean)
        .join('')})`,
      item.args.min ? `${mapper.min}(${item.args.min})` : undefined,
      item.args.max ? `${mapper.max}(${item.args.max})` : undefined,
    ]
      .filter(Boolean)
      .join('')
  }

  if (isKeyword(item, schemaKeywords.enum)) {
    if (item.args.asConst) {
      return `${zodKeywordMapper.union}([${item.args.items
        .map((item) => {
          return parseZodMeta(
            {
              keyword: schemaKeywords.const,
              args: item,
            },
            options,
          )
        })
        .join(', ')}])`
    }

    return `${value}([${item.args.items
      .map((item) => {
        if (item.format === 'number') {
          return transformers.stringify(item.value)
        }
        return transformers.stringify(item.value)
      })
      .join(', ')}])`
  }

  if (isKeyword(item, schemaKeywords.ref)) {
    return `${mapper.lazy}(() => ${item.args?.name})`
  }

  if (isKeyword(item, schemaKeywords.object)) {
    const properties = Object.entries(item.args?.properties || {})
      .filter((item) => {
        const schema = item[1]
        return schema && typeof schema.map === 'function'
      })
      .map((item) => {
        const name = item[0]
        const schema = item[1]

        return `"${name}": ${sort(schema)
          .map((item) => parseZodMeta(item, options))
          .filter(Boolean)
          .join('')}`
      })
      .join(',')

    const additionalProperties = item.args?.additionalProperties?.length
      ? item.args.additionalProperties
          .map((schema) => parseZodMeta(schema, options))
          .filter(Boolean)
          .at(0)
      : undefined

    const text = [
      `${value}({${properties}})`,
      item.args?.strict ? `${mapper.strict}()` : undefined,
      additionalProperties ? `${mapper.catchall}(${additionalProperties})` : undefined,
    ].filter(Boolean)

    return text.join('')
  }

  if (isKeyword(item, schemaKeywords.tuple)) {
    return `${value}([${sort(item.args)
      .map((arrayItem) => parseZodMeta(arrayItem, options))
      .filter(Boolean)
      .join(', ')}])`
  }

  if (isKeyword(item, schemaKeywords.const)) {
    if (item.args.format === 'number') {
      return `${value}(${transformers.toNumber(item.args.value)})`
    }
    return `${value}(${transformers.stringify(item.args.value)})`
  }

  if (isKeyword(item, schemaKeywords.matches)) {
    if (item.args) {
      return `${value}(${transformers.toRegExpString(item.args)})`
    }
  }

  if (isKeyword(item, schemaKeywords.default)) {
    if (item.args) {
      return `${value}(${item.args})`
    }
  }

  if (isKeyword(item, schemaKeywords.describe)) {
    if (item.args) {
      return `${value}(${transformers.stringify(item.args.toString())})`
    }
  }

  if (item.keyword in mapper && 'args' in item) {
    return `${value}(${(item as SchemaKeywordBase<unknown>).args as string})`
  }

  if (item.keyword in mapper) {
    return `${value}()`
  }

  return undefined
}
export function zodParser(schemas: Schema[], options: ParserOptions): string {
  if (!schemas.length) {
    return `export const ${options.name} = '';`
  }

  const sortedSchemas = sort(schemas)

  const JSDoc = createJSDocBlockText({
    comments: [options.description ? `@description ${transformers.jsStringEscape(options.description)}` : undefined].filter(Boolean),
  })

  const constName = `${JSDoc}\nexport const ${options.name}`
  const typeName = options.typeName ? ` as z.ZodType<${options.typeName}>` : ''
  const output = sortedSchemas
    .map((item) => parseZodMeta(item, options))
    .filter(Boolean)
    .join('')

  if (options.keysToOmit?.length) {
    const suffix = output.endsWith('.nullable()') ? '.unwrap().schema.and' : '.schema.and'
    const omitText = `${suffix}(z.object({ ${options.keysToOmit.map((key) => `${key}: z.never()`).join(',')} }))`
    return `${constName} = ${output}${omitText}${typeName}\n`
  }

  return `${constName} = ${output}${typeName}\n`

  // const root = createRoot()

  // if (!schemas.length) {
  //   root.render(
  //     <Text.Const
  //       export
  //       name={options.name}
  //       JSDoc={{ comments: [describeSchema ? `@description ${transformers.stringify(describeSchema.args)}` : undefined].filter(Boolean) }}
  //     >
  //       {'undefined'}
  //     </Text.Const>,
  //   )
  // } else {
  //   root.render(
  //     <Text.Const
  //       export
  //       name={options.name}
  //       JSDoc={{ comments: [describeSchema ? `@description ${transformers.stringify(describeSchema.args)}` : undefined].filter(Boolean) }}
  //     >
  //       {[
  //         sortedItems.map((item) => parseZodMeta(item, options)).filter(Boolean).join(''),
  //         options.keysToOmit?.length ? `.schema.and(z.object({ ${options.keysToOmit.map((key) => `${key}: z.never()`).join(',')} }))` : undefined,
  //         options.typeName ? ` as z.ZodType<${options.typeName}>` : '',
  //       ].filter(Boolean).join('') || ''}
  //     </Text.Const>,
  //   )
  // }

  // return root.output
}
