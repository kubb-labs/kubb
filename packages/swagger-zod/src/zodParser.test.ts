import { schemas } from '../../swagger/mocks/schemas.ts'
import { parseZodMeta, zodParser } from './zodParser.tsx'

describe('parseZodMeta', () => {
  test.each(schemas.basic)('$name', ({ name, schema }) => {
    const text = parseZodMeta(schema, { name })
    expect(text).toMatchSnapshot()
  })
})

describe('zodParser', () => {
  test.each(schemas.full)('$name', ({ schema, name }) => {
    const text = zodParser(schema, { name })
    expect(text).toMatchSnapshot()
  })
})
