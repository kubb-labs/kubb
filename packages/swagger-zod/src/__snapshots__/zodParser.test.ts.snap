// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseZodMeta > 'and' 1`] = `"z.string().and(z.number())"`;

exports[`parseZodMeta > 'any' 1`] = `"z.any()"`;

exports[`parseZodMeta > 'array' 1`] = `"z.array(z.union([z.string(), z.number()]))"`;

exports[`parseZodMeta > 'arrayAdvanced' 1`] = `"z.array(z.number().min(1)).min(3).max(10)"`;

exports[`parseZodMeta > 'arrayEmpty' 1`] = `"z.array()"`;

exports[`parseZodMeta > 'arrayRef' 1`] = `"z.array(z.lazy(() => Pet))"`;

exports[`parseZodMeta > 'blob' 1`] = `undefined`;

exports[`parseZodMeta > 'boolean' 1`] = `"z.boolean()"`;

exports[`parseZodMeta > 'catchall' 1`] = `"z.object({}).catchall(z.lazy(() => Pet))"`;

exports[`parseZodMeta > 'date' 1`] = `"z.date()"`;

exports[`parseZodMeta > 'datetime' 1`] = `"z.string().datetime()"`;

exports[`parseZodMeta > 'default' 1`] = `".default()"`;

exports[`parseZodMeta > 'default' 2`] = `".default(default)"`;

exports[`parseZodMeta > 'enum' 1`] = `"z.enum(["A", "B", "C", "2"])"`;

exports[`parseZodMeta > 'matches' 1`] = `".regex(new RegExp('^[A-Z]{2}$'))"`;

exports[`parseZodMeta > 'matchesReg' 1`] = `".regex(new RegExp('node_modules'))"`;

exports[`parseZodMeta > 'max' 1`] = `".max(2)"`;

exports[`parseZodMeta > 'min' 1`] = `".min(2)"`;

exports[`parseZodMeta > 'nullable' 1`] = `".nullable()"`;

exports[`parseZodMeta > 'number' 1`] = `"z.number()"`;

exports[`parseZodMeta > 'object' 1`] = `"z.object({"firstName": z.string().min(2),"address": z.string().describe("Your address").nullable()})"`;

exports[`parseZodMeta > 'objectAnd' 1`] = `"z.object({"firstName": z.string().min(2).default(test),"age": z.number().min(2).default(2),"address": z.string().and(z.number()).describe("Your address").nullable()})"`;

exports[`parseZodMeta > 'objectEmpty' 1`] = `"z.object({})"`;

exports[`parseZodMeta > 'objectOptional' 1`] = `"z.object({"firstName": z.string().min(2).optional(),"address": z.string().describe("Your address").nullable()})"`;

exports[`parseZodMeta > 'ref' 1`] = `"z.lazy(() => Pet)"`;

exports[`parseZodMeta > 'string' 1`] = `"z.string()"`;

exports[`parseZodMeta > 'tuple' 1`] = `"z.tuple([])"`;

exports[`parseZodMeta > 'tupleMulti' 1`] = `"z.tuple([z.string(), z.number()])"`;

exports[`parseZodMeta > 'undefined' 1`] = `"z.undefined()"`;

exports[`parseZodMeta > 'union' 1`] = `"z.union([z.string(), z.number()])"`;

exports[`parseZodMeta > 'unionOne' 1`] = `"z.string()"`;

exports[`parseZodMeta > 'unknown' 1`] = `"z.unknown()"`;

exports[`zodParser > 'Object' 1`] = `
"
export const Object = z.object({"firstName": z.string().min(2).default(test),"age": z.number().min(2).default(2),"address": z.string().and(z.number()).describe("Your address").nullable()}).describe("Your address").nullable()
"
`;

exports[`zodParser > 'Order' 1`] = `
"
export const Order = z.object({"status": z.enum(["placed", "approved"])})
"
`;

exports[`zodParser > 'Record' 1`] = `
"
export const Record = z.object({}).catchall(z.number()).optional()
"
`;

exports[`zodParser > 'Upload' 1`] = `
"
export const Upload = 
"
`;
